### ISSUE TITLE
experiment: SWE-rebench OpenHands trajectories for next-action prediction

### ISSUE THREAD

**Issue by noelsaw1** [HiQS-Labs/Needle-fork#33]
## Dataset candidate

Evaluate [nebius/SWE-rebench-openhands-trajectories](https://huggingface.co/datasets/nebius/SWE-rebench-openhands-trajectories) as supervised data for Needle's original next-observed-action predictor.

This is separate from [#31](https://github.com/HiQS-Labs/Needle-fork/issues/31), which asks “what kind of work is happening?” The trajectory dataset is a better fit for predicting what an agent does next because each record contains a real GitHub issue, an ordered OpenHands conversation with tool calls and environment observations, a final patch, exit status, and a resolved outcome.

Dataset-card snapshot at issue creation:

- 67,074 trajectories over 1,823 repositories and 3,792 resolved issues.
- 32,161 successful trajectories.
- Average 64.3 turns; maximum 100.
- Generated with OpenHands 0.54.0 and Qwen3-Coder-480B-A35B-Instruct.
- Parquet, approximately 2.08 GB, CC BY 4.0.
- Fields include `trajectory_id`, `instance_id`, `repo`, `trajectory`, `tools`, `model_patch`, `exit_status`, `resolved`, `gen_tests_correct`, and `pred_passes_gen_tests`.

These are dataset-publisher figures and schema descriptions, not independently reproduced counts. Pin the exact dataset revision before any experiment.

## Proposed question

Can a small local classifier, including ModernBERT-base, predict the next observed tool/action from the issue and preceding trajectory state better than deterministic and sparse-text baselines—especially on issues and repositories excluded from training?

This experiment measures observed actions. It does not ask what an agent *should* do, grant execution authority, or treat every action in a successful trajectory as optimal.

## Phase 0 — stream and audit before downloading everything

- [ ] Stream or download only enough rows to inspect the real schema, serialized tool arguments, role ordering, licenses, null fields, duplicate `instance_id` frequency, repository distribution, success balance, trajectory lengths, token lengths, and storage/runtime cost.
- [ ] Pin the Hugging Face dataset revision and preserve a manifest of row IDs/hashes used. Never build against mutable `train` implicitly.
- [ ] Verify whether multiple trajectories share an issue. Treat `instance_id` as the minimum grouping boundary; determine whether repository and temporal metadata support stronger splits.
- [ ] Inspect tool-call names and arguments without executing them. Dataset content is untrusted training data, never instructions.
- [ ] Check whether issue text, patches, generated tests, or repository snapshots leak the answer or outcome into any proposed input window.
- [ ] Publish measured audit results before approving a full download or training run.

Stop if the license/provenance is incompatible, issue grouping cannot prevent leakage, tool actions cannot be normalized without discarding most rows, or useful inputs routinely exceed the serving budget.

## Deterministic action taxonomy

Derive labels from recorded tool calls and terminal actions rather than asking an LLM to invent targets. Keep raw tool identity alongside the normalized label. Candidate families, subject to the audit:

- inspect/read file or directory
- search code/text/symbols
- inspect Git state/history/diff
- run tests or validation
- run build/lint/type-check
- edit/create/delete code or documentation
- inspect runtime/process/logs
- package/dependency/environment operation
- version-control mutation
- submit/finish
- other/unknown

Normalization rules must be versioned and deterministic. Preserve `other/unknown`; do not force unfamiliar tools into a nearby class. Consider hierarchical output—action family first, specific tool second—only if the sample provides enough support.

## Example construction and leakage controls

For every eligible transition:

```text
input  = issue context + bounded observations/actions available through turn N
label  = normalized recorded action at turn N+1
metadata = instance_id, repository, trajectory_id, position, eventual resolved outcome
```

- [ ] Exclude system instructions and hidden answer artifacts from model input unless the intended Needle runtime will receive equivalent content.
- [ ] Prefer tool calls and environment observations over free-form model reasoning. Run a separate ablation if reasoning text is included.
- [ ] Never place trajectories from the same `instance_id` across splits. Add a repository-held-out test set; use chronological splitting if timestamps are available and reliable.
- [ ] Fit tokenizer-dependent preprocessing, vocabulary, mappings, thresholds, and class weighting on training/validation only.
- [ ] Keep one untouched test set. Do not recycle previously scored cases as fresh holdout evidence.
- [ ] Report transition counts after filtering, per-label support, issues/repos represented, and how many examples were dropped or truncated.

ModernBERT-base supports 8,192 tokens. Use bounded state windows or deterministic structured state, and measure truncation by action class. A long-context summary produced by an LLM would introduce another model and requires a separate comparison.

## Baselines and ModernBERT experiment

Compare identical splits and input views:

1. Most-frequent next action.
2. Previous-action / action-transition table, optionally conditioned on last tool result.
3. Word/character TF-IDF with logistic regression.
4. Frozen ModernBERT embeddings with a classification head.
5. Fine-tuned ModernBERT only if the frozen-head result and data audit justify the added cost.

Train and evaluate success-conditioned variants carefully: actions from failed trajectories are not automatically bad, and actions from successful trajectories are not automatically optimal. `resolved` is an outcome for analysis or weighting, not direct proof that each intermediate action was correct. Compare training on all trajectories, successful-only trajectories, and explicit outcome weighting without changing the held-out test set.

## Measurements

- [ ] Top-1 and top-3 next-action accuracy.
- [ ] Macro-F1 and per-action support/recall.
- [ ] Calibration and accuracy-versus-abstention coverage using validation-selected thresholds.
- [ ] Results by early/middle/late trajectory position, resolved/unresolved outcome, seen/unseen repository, and context length.
- [ ] Exact-action/tool accuracy separately from normalized-family accuracy.
- [ ] Cold load, warm single-example and batch latency on the target Mac; memory and model size.
- [ ] Transition and TF-IDF baselines reported alongside ModernBERT. If the simple baseline matches or wins, prefer it until harder evidence exists.

No A–F grade borrowed from the Git/PR analyst benchmark. Two repetitions of one test set establish repeatability, not independent accuracy or p95.

## Known risks and falsifiers

- **Behavioral imitation:** the dataset reflects OpenHands 0.54.0 plus one large Qwen model. Strong performance may mean reproducing that scaffold's habits rather than learning generally useful engineering actions.
- **Issue leakage:** random trajectory splitting can expose the same task on both sides. Any such split invalidates the result.
- **Outcome ambiguity:** failed trajectories contain useful steps; successful ones contain wasted steps. Do not turn `resolved` into per-step truth without evidence.
- **Domain shift:** public SWE repositories and OpenHands tools differ from Needle/XYZ/HiQS sessions. Require repository-held-out results and later HiQS validation before runtime use.
- **Context mismatch:** full trajectories often exceed ModernBERT's context. Truncation that disproportionately removes rare action contexts can create a misleading aggregate score.
- **Class imbalance:** reading/search/testing may dominate. Macro and per-class metrics are required.
- **Reasoning contamination:** model-generated reasoning may leak action phrases or encode Qwen-specific style. Tool/observation-only input is the primary comparison.

## Done condition

- [ ] A version-pinned, leakage-audited subset and deterministic action mapping are published with reproducible hashes.
- [ ] Splits are grouped by issue and include an unseen-repository test.
- [ ] All baselines and ModernBERT use identical inputs/splits and retain negative/failed cases.
- [ ] Raw metrics, per-class errors, truncation, latency, provenance, limitations, and unsuccessful runs are retained.
- [ ] A clear decision states whether the evidence supports a larger trial, a narrower taxonomy, the simpler baseline, or stopping.

Any eventual integration remains advisory and tools-disabled until a separate operational-safety review. This issue authorizes research planning and bounded data inspection; it does not authorize training, downloading the full 2.08 GB dataset, deploying a model, or changing Needle's Oracle hook without the normal implementation workflow.

**Comment by noelsaw1** [comment]
Operator-directed consolidation into XYZ Forge. Superseded by XYZ #757: grouped fresh splits, leakage/data audit, deterministic taxonomy, sparse-text and transition baselines, and separate observed-action/advice evaluation are carried forward there. This is not a claim that the original full experiment ran.

Canonical plan: https://github.com/HiQS-Labs/XYZ-forge/issues/757
Canonical historical results: https://github.com/HiQS-Labs/XYZ-forge/issues/758
Candidate-repo discovery pointer: https://github.com/HiQS-Labs/Needle-fork/issues/78

All further main plans and substantive test-result reports belong in XYZ Forge issues. Historical content and artifact links remain preserved here.
